torch/distributed/_state_dict_utils.py

```diff
@@ -457,3 +457,93 @@ def tensor_func(
         return False
 
     return True
+
+
+class _TensorInfo(NamedTuple):
+    size: torch.Size
+    dtype: torch.dtype
+
+
+def _broadcast_tensors(
+    full_state_dict: Dict[str, Any],
+    local_state_dict: Dict[str, Any],
+    keys: List[str],
+    device: torch.device,
+    pg: Optional[dist.ProcessGroup] = None,
+) -> None:
+    tensors = []
+    for key in keys:
+        if dist.get_rank() == 0:
+            full_state = full_state_dict[key]
+            assert isinstance(full_state, torch.Tensor)
+            full_tensor = full_state.detach().to(device)
+        else:
+            tensor_info = full_state_dict[key]
+            full_tensor = torch.empty(
+                size=tensor_info.size,
+                device=device,
+                dtype=tensor_info.dtype,
+            )
+
+        tensors.append(full_tensor)
+        local_state = local_state_dict.get(key, None)
+        if local_state is None:
+            continue
+        elif isinstance(local_state, DTensor):
+            local_state_dict[key] = (local_state, full_tensor)
+        else:
+            local_state_dict[key] = full_tensor
+
+    if pg is None:
+        pg = dist.distributed_c10d._get_default_group()
+    dist._broadcast_coalesced(pg, tensors, 500, 0)
```

Comment:
After this `_broadcast_coalesced` call, does every rank have the full state dict in GPU memory?

I think no matter what, we want to interleave the broadcast with sharding. Otherwise, we either use 8x CPU memory (not `mmap`'d) across the host or 1x GPU memory per GPU.

Reply:
Right, I forgot to do overlapping. The new version already change it. But do you still recommend avoid using broadcast_coalesced and just use the regular broadcast?